tests/nn/data_parallel/test_fsdp_regnet.py

```diff
@@ -9,86 +9,216 @@
+            inputs[rank].append(torch.rand(2, 2, 2, 2))
+
+    # Run DDP training twice, fp and mp.
+    for precision in ["full", "mixed"]:
+        temp_file_name = tempfile.mkstemp()[1]
+        unused = tempfile.mkstemp()[1]
+        rank_0_output = tempfile.mkstemp()[1]
+        try:
+            fsdp_config = None  # This means we use DDP in _test_func.
+            mp.spawn(
+                _test_func,
+                args=(
+                    world_size,
+                    fsdp_config,
+                    precision == "mixed",
+                    temp_file_name,
+                    unused,
+                    state_before,
+                    inputs,
+                    rank_0_output,
+                    None,
+                ),
+                nprocs=world_size,
+                join=True,
+            )
+            if precision == "full":
+                state_after_fp = torch.load(rank_0_output)
+            else:
+                state_after_mp = torch.load(rank_0_output)
+        finally:
+            rmf(temp_file_name)
+            rmf(unused)
+            rmf(rank_0_output)
+
+    assert state_dict_norm(state_after_fp) != state_dict_norm(state_after_mp)
+
+    return state_before, inputs, state_after_fp, state_after_mp
+
+
+# A fixture to get tempfiles and ensure they are cleaned up.
+@pytest.fixture()
+def temp_files():
+    temp_file_name = tempfile.mkstemp()[1]
+    unused = tempfile.mkstemp()[1]
 
-def _test_func(rank, world_size, fsdp_config, tempfile_name, unused):
+    yield temp_file_name, unused
+
+    # temp files could have been removed, so we use rmf.
+    rmf(temp_file_name)
+    rmf(unused)
+
+
+def _test_func(
+    rank,
+    world_size,
+    fsdp_config,
+    ddp_mixed_precision,
+    tempfile_name,
+    unused,
+    state_before,
+    inputs,
+    rank_0_output,
+    state_after,
+):
     result = dist_init(rank, world_size, tempfile_name, unused)
     assert result, "Dist init failed"
 
-    assert isinstance(fsdp_config, dict), str(fsdp_config)
-
-    class Model(Module):
-        def __init__(self):
-            super().__init__()
-            # TODO (Min): for now, we just test pytorch sync_bn here.
-            #             this will grow into regnet; testing apex sync_bn, etc.
-            self.conv = Conv2d(2, 2, (1, 1))
-            self.bn = BatchNorm2d(2)
-
-        def forward(self, x):
-            x = self.conv(x)
-            x = self.bn(x)
-            return x
-
-    # TODO (Min): check DDP equivalency.
+    ddp = True
+    if fsdp_config:
+        ddp = False
+        assert isinstance(fsdp_config, dict), str(fsdp_config)
 
     model = Model()
-    # Note, different rank may wrap in different order due to different random
-    # seeds. But results should be the same.
-    if random.randint(0, 1) == 0:
-        print("auto_wrap_bn, then convert_sync_batchnorm")
-        model = auto_wrap_bn(model)
+    model.load_state_dict(state_before)
+    model = model.cuda()
```

Comment:
probably me but I'm not seeing where the unit test is guarded for cuda ?

Reply:
it is on line 200, `skip_if_single_gpu`.  I think it really meant single or no GPU. :-)